//MKLINK timestamp.proto C:\code\libraries\protobuf\src\google\protobuf\timestamp.proto
//ln -s $REPO_DIR/protobuf/src/google/protobuf/timestamp.proto .
//protoc --cpp_out dllexport_decl=Γ:. -I$REPO_DIR/protobuf/src/ -I. messages.proto
//%REPO_DIR%\protobuf\cmake\build\debug\protoc --cpp_out dllexport_decl=Γ:. -I%REPO_DIR%/protobuf/src/ -I. messages.proto
//npx pbjs -t static-module -w es6 -o src/app/proto/messages.js messages.proto & npx pbts -o src/app/proto/messages.d.ts src/app/proto/messages.js

syntax = "proto3";

package Jde.Logging.Proto;
import "google/protobuf/timestamp.proto";

enum ELogLevel
{
	Trace = 0;
	Debug = 1;
	Information = 2;
	Warning = 3;
	Error = 4;
	Critical = 5;
	LogLevelNone = 6;
	NoLog = -1;
}

message Instance
{
	string application=1;
	string host=2;
	uint32 pid=3;
	ELogLevel server_log_level=4;
	ELogLevel client_log_level=5;
	uint32 start_time=6;
	uint32 rest_port=7;
	uint32 websocket_port=8;
	string instance_name=9;
}

message Message
{
	google.protobuf.Timestamp Time=1;
	ELogLevel Level=2;
	uint32 MessageId=3;
	uint32 FileId=4;
	uint32 FunctionId=5;
	uint32 LineNumber=6;
	uint32 UserId=7;
	uint64 ThreadId=8;
	repeated string Variables=9;
}

message SessionInfo
{
	google.protobuf.Timestamp expiration=1;
	uint32 session_id=2;
	uint32 user_id=3;
};

enum EFields
{
	FieldsNone=0;
	TimestampField=0x1;
	MessageId=0x2;
	MessageField=0x4;
	Level=0x8;
	FileId=0x10;
	File=0x20;
	FunctionId=0x40;
	Function=0x80;
	LineNumber=0x100;
	UserId=0x200;
	User=0x400;
	ThreadId=0x800;
	Thread=0x1000;
	VariableCount=0x2000;
	SessionId=0x4000;
}

message RequestString
{
	EFields Field=1;
	uint32 Id=2;
	string Value=3;
}

message Status
{
	uint32 StartTime=1;
	uint64 Memory=2;
	ELogLevel ServerLogLevel=3;
	ELogLevel ClientLogLevel=4;
	repeated string Details=5;
}

message CustomMessage
{
	uint32 RequestId=1;
	bytes Message=2;
}

message CustomComplete
{
	uint32 RequestId=1;
}

message RequestSessionInfo
{
	uint32 session_id=1;
}

message ToServerUnion
{
	oneof Value
	{
		Instance Instance = 1;
		Message Message = 2;
		RequestString String = 3;
		Status Status = 4;
		CustomMessage Custom=5;
		CustomComplete Complete=6;
		RequestSessionInfo session_info=7;
	}
}

message ToServer
{
	repeated ToServerUnion Messages = 1;
}

//////////////////////////////////////////////////////////////////////////////////////////////////////////

enum EFromServer
{
	Ping=0;
	Error_=1;
	ReadError=2;
}

message Acknowledgement
{
	uint32 InstanceId=1;
}

message GenericFromServer
{
	EFromServer What = 1;
	string Description=2;
}

message Strings
{
	repeated uint32 Messages = 1;
	repeated uint32 Files = 2;
	repeated uint32 Functions = 3;
	repeated uint32 Threads = 4;
}

message Progress
{
	uint32 RequestId=1;
	double Value=2;
}
message LogLevels
{
	ELogLevel Server=1;
	ELogLevel Client=2;
}

message FromServerUnion
{
	oneof Value
	{
		Acknowledgement Acknowledgement = 1;
		GenericFromServer Generic = 2;
		Strings Strings = 3;
		LogLevels LogLevels = 4;
		CustomMessage Custom = 5;
		Progress Progress = 6;
		SessionInfo session_info = 7;
	}
}

message FromServer
{
	repeated FromServerUnion Messages = 1;
}